### ISSUE TITLE
stagnierender Trainingsloss und Accuracy bei ConvSegment Modell

### ISSUE THREAD

**Issue by hannahweber244** [hannahweber244/CryptoSentiment#4]
- Optimierung stagniert schon in den ersten Epochen, sowohl Accuracy, als auch Trainingsloss werden nicht signifikant besser
Es wurden bereits ausprobiert:
- unterschiedliche Lossfunktionen
- unterschiedliche Lernraten

Issue bezieht sich auf Conv/Rnn-Sentiment.ipynb

**Comment by hannahweber244** [comment]
![image](https://user-images.githubusercontent.com/58035222/123510920-7c072b00-d67e-11eb-9179-0af74f946faf.png)

Beispielhafter Screenshot

**Comment by cbiehl** [comment]
Hi, hier eine paar Hinweise von mir:
- Die Berechnung der Accuracy in der `test()` Funktion scheint innerhalb des Batches nicht jede einzelne Instanz zu vergleichen, sondern nur accuracy += 1 auf Basis des argmax über den ganzen Batch auszuführen. Das würde erklären warum sich der Wert auch nie ändert.

- Bei der Loss Funktion bitte darauf achten was ihr in der forward() Methode macht, Beispiel: nn.NLLLoss erwartet dass die nn.LogSoftmax (oder F.log_softmax) angewandt wurde, aber nn.CrossEntropyLoss erwartet dass keine Softmax Aktivierung genutzt wird (beim Mean Squared Error natürlich auch, war überrascht diesen auch im Code zu sehen, zumal die Labels als One-Hot Vektoren geladen wurden).

- Startet am besten mit einem einfachen Modell und baut es dann aus wenn es funktioniert.

- Nutzt einen Debugger (z.B. in PyCharm, Visual Studio Code oder Spyder) und geht durch jede einzelne Zeile. Achtet auf die Shapes der Tensoren / Arrays.

- Vielleicht hilft es euch für eure Experimente kein Fine-Tuning der Embeddings durchzuführen, dadurch dauert der Trainingsprozess sehr viel länger (z.B. durch Initialisierung mit nn.Embedding.from_pretrained(matrix, freeze=True)

- Testet eine höhere Lernrate (aber Vorsicht: eine zu hohe Lernrate kann zu "Dying RELUs" führen, i.e. Parametern durch die die ReLU immer 0 ausgeben würde)

- Ich kann diesen Guide empfehlen: http://karpathy.github.io/2019/04/25/recipe/

- Sprecht mich gerne am Donnerstag nochmal darauf an

**Comment by hannahweber244** [comment]
- Training ohne Mitoptimierung der Embeddings halbiert die Trainingszeit pro Epoche
- Vereinfachung des Netzes resultiert in um 5% reduzierte Accuracy --> Modell lernt immer noch nicht
- torch.argmax ist in dem vorliegenden Fall kein Problem, da die Testfunktion mit einer batch_size = 1 aufgerufen wird
- Ausgabe der Conv Layer zeigt, dass nach jedem Schritt die Werte kleiner werden und schlussendlich gegen 0 streben! (Lossfunktion MSE und NLLLoss) --> dying ReLU Problem 
- Auch mit Tanh oder Leaky ReLU bleibt 0en Problem nach Convolution bestehen --> Trainingsloss wird mit jeder Epoche nur minimal verbessert
- ohne eine Aktivierungsfunktion gleiches Problem

**Comment by hannahweber244** [comment]
Versuch der input-indipendent-baseline zeigt:
- wenn nur 0en ins Netz zum Training gegeben werden und auf "echten" Daten getestet wird, werden exakt gleiche Ergebnisse erreicht, wie auf den realen Trainingsdaten --> Modell lernt keine Informationen aus den Trainingsdaten
- padding_token = 0 und große Sequenzlänge (871) sorgen dafür, dass große Teile einer jeden Trainingsdateninstanz 0 wird
- auch verkürzen der Inputsequenzläne (100 und 50) führt zu keiner Verbesserung

**Comment by hannahweber244** [comment]
Weitere Lösungsansätze:
- Layergewichte selbst auf 1 initialisieren (deutlich höher als die random initialisierten Gewichte von PyTorch)
Ergebnis:
- Trainingsloss lässt sich in den ersten ca. 3 Epochen verbessern, stagniert aber bei bereits bekanntem Loss, genauso wie Accuracy

**Comment by hannahweber244** [comment]
- gleiche Problematik tritt bei einer LSTM basierten Implementierung auf: Gleiche Accuracy wie bei Conv MOdell und kein Lerneffekt über die Epochen hinweg --> Problem scheint in der Verarbeitung der Embeddings, bzw. in den Daten selbst zu liegen?

**Comment by Julian-Greil** [comment]
![image](https://user-images.githubusercontent.com/58034641/124127019-8a3cb900-da7b-11eb-8442-ceb9975da431.png)

- https://www.kaggle.com/ankurzing/sentiment-analysis-for-financial-news
- Mithilfe des obigen Datensatzes das selbe Netzwerk trainiert und überraschend gute Ergebnisse bekommen? Möglicherweise zu gut? Genaueres betrachten nochmal notwendig.

**Comment by hannahweber244** [comment]
- fehlerhafte Benennung einer Variable in Listcomprehension --> danach wird Accuracy von 89%
